src/ssh.rs

```diff
@@ -52,17 +53,37 @@ impl Session {
         let mut sess = ssh2::Session::new().ok_or(Context::new("libssh2 not available"))?;
         sess.handshake(&tcp)
             .context("failed to perform ssh handshake")?;
-        sess.userauth_pubkey_file(username, None, key, None)
-            .context("failed to authenticate ssh session")?;
+        if let Some(key) = key {
+            sess.userauth_pubkey_file(username, None, key, None)
+                .context("failed to authenticate ssh session with key")?;
+        } else {
+            // TODO doesn't work, see crate::providers::baremetal::test
+            let mut ag = sess.agent()?;
+            ag.connect().context("could not connect to ssh-agent")?;
+            ag.list_identities()
+                .context("could not list ssh-agent identities")?;
+            let ok = ag.identities().flat_map(|x| x).any(|id| {
+                ag.userauth(username, &id)
+                    .map_err(|e| {
+                        trace!(log, "agent identity failed"; "username" => username,  "identity" => id.comment(), "err" => ?e);
+                        e
+                    })
+                    .is_ok()
+            });
+            if !ok {
+                bail!("failed to authenticate ssh session with ssh-agent");
+            }
+        }
 
         Ok(Session {
+            addr,
             ssh: sess,
             _stream: tcp,
         })
     }
 
     /// Issue the given command and return the command's raw standard output.
```

Comment:
This should be updated to reflect what it now returns. Maybe it should return a `struct` rather than a tuple?